File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.rb
# @param {String} s
# @return {Integer}
def length_of_longest_substring(s)
    map = Hash.new(0)
    start = 0
    numOver = 0
    len = 0
    for right in 0..s.length-1
        map[s[right]] += 1
        if(map[s[right]] == 2) 
            numOver += 1
        end
        while(numOver > 0)
            if(map[s[start]] == 2)
                numOver -= 1
            end
            map[s[start]] -= 1
            start += 1
        end
        len = [len, right - start + 1].max
    end
    len
end

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int[] map = new int[128]; //Every ascii value
        int len = 0;
        int start = 0;
        for(int end = 0; end < s.length(); end++){
            map[s.charAt(end)] += 1;
            while(!validMap(map)){ 
                map[s.charAt(start)] -= 1;
                start++;
            }
            len = Math.max(len, end - start + 1);
        }
        return len;
    }
    private boolean validMap(int[] map){
        for(int i = 0; i < map.length; i++){
            if(map[i] > 1){
                return false;
            }
        }
        return true;
    }
}